File: kafka/src/test/groovy/io/micronaut/configuration/kafka/docs/consumer/offsets/ack/ProductListener.java
package io.micronaut.configuration.kafka.docs.consumer.offsets.ack;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.configuration.kafka.docs.consumer.config.Product;
import io.micronaut.messaging.Acknowledgement;
// end::imports[]

import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST;
import static io.micronaut.configuration.kafka.annotation.OffsetStrategy.DISABLED;

class ProductListener {

    // tag::method[]
    @KafkaListener(offsetReset = EARLIEST, offsetStrategy = DISABLED) // <1>
    @Topic("awesome-products")
    void receive(Product product,
                 Acknowledgement acknowledgement) { // <2>

        // process product record

        acknowledgement.ack(); // <3>
    }
    // end::method[]
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/scope/ProductListener.py
# tag::imports[]
import uuid

from micronaut.configuration.kafka.annotation import KafkaListener, KafkaScope, OffsetReset, Topic
# end::imports[]


# tag::scope[]
@KafkaScope
class ProductMetadata:

    def __init__(self):
        self.correlation_id = str(uuid.uuid4())

    def get_correlation_id(self) -> str:
        return self.correlation_id
# end::scope[]


# tag::listener[]
@KafkaListener(offsetReset=OffsetReset.EARLIEST)
class ProductListener:

    def __init__(self, product_metadata: ProductMetadata):
        self.product_metadata = product_metadata

    @Topic("products")
    def receive(self, product: str) -> None:
        print("Received " + product + " with correlation " + self.product_metadata.get_correlation_id())
# end::listener[]
